browser/speedreader/speedreader_tab_helper.cc

```diff
@@ -33,15 +35,52 @@ bool SpeedreaderTabHelper::IsSpeedreaderEnabled() const {
   return SpeedreaderServiceFactory::GetForProfile(profile)->IsEnabled();
 }
 
+bool SpeedreaderTabHelper::IsEnabledForSite() {
+  if (!IsSpeedreaderEnabled())
+    return false;
+
+  Profile* profile =
+      Profile::FromBrowserContext(web_contents()->GetBrowserContext());
+  auto* content_rules = HostContentSettingsMapFactory::GetForProfile(profile);
+  return speedreader::IsEnabledForSite(content_rules,
+                                       web_contents()->GetLastCommittedURL());
+}
+
+void SpeedreaderTabHelper::MaybeToggleEnabledForSite(bool on) {
+  if (!IsSpeedreaderEnabled())
+    return;
+
+  Profile* profile =
+      Profile::FromBrowserContext(web_contents()->GetBrowserContext());
+  auto* content_rules = HostContentSettingsMapFactory::GetForProfile(profile);
+  bool enabled = speedreader::IsEnabledForSite(
+      content_rules, web_contents()->GetLastCommittedURL());
+  if (enabled != on) {
+    speedreader::SetEnabledForSite(content_rules,
+                                   web_contents()->GetLastCommittedURL(), on);
+    web_contents()->GetController().Reload(content::ReloadType::NORMAL, false);
+  }
+}
+
+void SpeedreaderTabHelper::SingleShotSpeedreader() {
+  single_shot_next_request_ = true;
+  auto* contents = web_contents();
+  if (contents)
+    contents->GetController().Reload(content::ReloadType::NORMAL, false);
+}
+
 void SpeedreaderTabHelper::UpdateActiveState(
     content::NavigationHandle* handle) {
   DCHECK(handle);
   DCHECK(handle->IsInMainFrame());
 
-  const bool enabled = IsSpeedreaderEnabled();
+  if (single_shot_next_request_) {
+    distill_state_ = DistillState::kReaderMode;
```

Comment:
I would clear the flag right there, not 100% sure DidStopLoading is reliable

Reply:
done